megatron/core/extensions/gemm_patch.py

```diff
@@ -0,0 +1,187 @@
+            if not same_cksum:
+                print(f"RANK {get_rank_info()}: DIFFERENT CHECKSUM ON 2ND RUN for dtype {get_dtype(A)}. {cksum[1]} != {cksum[0]}, "
+                      f"shapes A:{list(A.size())},B:{list(B.size())},out:{list(out_res.shape)}")
+            elif not torch.equal(rerun_res, out_res):
+                print(f"RANK {get_rank_info()}: DIFFERENT BITWISE on 2ND RUN for dtype {get_dtype(A)}. "
+                      f"shapes A:{list(A.size())},B:{list(B.size())},out:{list(out_res.shape)}")
+            del rerun_res
+        return res
+
+def new_general_grouped_gemm(
+    A,
+    B,
+    out,
+    out_dtype,
+    workspaces,
+    layout = "TN",
+    m_splits = None,
+    gelu = False,
+    grad =False,
+    accumulate = False,
+    bias = None,
+    use_bias = False,
+    use_split_accumulator = False,
+    D_dtype = None,
+    single_output = False,
+):
+    saved_output = None
+    out_will_rerun = None
+    bias_will_rerun = None
+    for run_idx in range(2):
+        is_rerun = run_idx == 1
+        if not is_rerun:
+            out_will_rerun = [t.clone().detach() for t in out] if out is not None and (accumulate or out is bias) else out
+            bias_will_rerun = out_for_rerun if bias is not None and out is bias else bias
+        res = orig_general_grouped_gemm(
+            A,
+            B,
+            out_will_rerun if not is_rerun else out,
+            out_dtype,
+            workspaces,
+            layout,
+            m_splits,
+            gelu,
+            grad if is_rerun else False,
+            accumulate,
+            bias_will_rerun if not is_rerun else bias,
+            use_bias,
+            use_split_accumulator,
+            D_dtype,
+            single_output,
+        )
+        output = res[0]
+        if not is_rerun:
+            saved_output = [t.clone().detach() for t in output]
+            continue
+        if is_rerun:
+            for tidx in range(len(output)):
+                if not torch.equal(output[tidx], saved_output[tidx]):
+                    print(
+                        f"RANK {get_rank_info()}: DIFFERENT GROUPED_GEMM RESULT ON 2ND RUN "
+                        f"FOR OP GroupedGEMM group {tidx} = {get_dtype(A[tidx])}({list(A[tidx].size())}) x {get_dtype(B[tidx])}({list(B[tidx].size())}) -> {get_dtype(output[tidx])}({list(output[tidx].size())})"
+                    )
+        return res
+
+def patch_te_gemms() -> None:
+    global _GLOBAL_IS_PATCHED_
+    if _GLOBAL_IS_PATCHED_ is not None and _GLOBAL_IS_PATCHED_:
+        return
+    transformer_engine.pytorch.cpp_extensions.gemm.general_gemm = new_general_gemm
+    transformer_engine.pytorch.cpp_extensions.gemm.general_grouped_gemm = new_general_grouped_gemm
+
+    rank = int(os.environ.get("RANK", 0))
+    for module in sys.modules:
+        if 'general_gemm' in dir(sys.modules[module]):
+            if rank == 0:
+                print(f"PATCHING general_gemm IN MODULE: {module}")
```

Comment:
No `print`s in `megatron/core`. Use Python's `logging` module instead.

Reply:
You can use https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/blob/main/megatron/core/utils.py?ref_type=heads#L737 to just log on a single rank.